File: ConstructBinaryTreeFromPreorderAndInorderTraversal.py
"""
105. Construct Binary Tree from Preorder and Inorder Traversal
Given preorder and inorder traversal of a tree, construct the binary tree.

Note:
You may assume that duplicates do not exist in the tree.
"""
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# O(n)
# O(n)
class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        lookup = {}
        for i in range(len(inorder)):
            lookup[inorder[i]] = i
        root = self.build(lookup, preorder, 0, len(preorder) - 1)
        return root
        
    def build(self, lookup, pre, in_start, in_end):
        if not pre:
            return
        node = TreeNode(pre[0])
        mid = lookup[pre[0]]
        length_left = mid - in_start
        node.left = self.build(lookup, pre[1: 1 + length_left], in_start, mid - 1)
        node.right = self.build(lookup, pre[1 + length_left: in_end + 1], mid + 1, in_end)
        return node

if __name__ == "__main__":
    root = Solution().buildTree([1, 2], [2, 1])

File: Java/ConstructBinaryTreefromPreorderandInorderTraversal.java
import java.util.*;

public class ConstructBinaryTreefromPreorderandInorderTraversal {
	
    public static TreeNode buildTree(int[] preorder, int[] inorder) {
        Map<Integer, Integer> lookup = new HashMap<Integer, Integer>();
        int inStart = 0;
        int preStart = 0;
        int inEnd = inorder.length - 1;
        for(int i = 0; i < inorder.length; i++) {
            lookup.put(inorder[i], i);
        }
        return helper(lookup, preorder, inStart, inEnd, preStart);
    }
    
    public static TreeNode helper(Map<Integer, Integer> lookup, int[] preorder, int inStart, int inEnd, int preStart) {
        if(inStart > inEnd) {return null;}
        TreeNode node = new TreeNode(preorder[preStart]);
        int position = lookup.get(preorder[preStart]);
        TreeNode left = helper(lookup, preorder, inStart, position - 1, preStart + 1);
        TreeNode right = helper(lookup, preorder, position + 1, inEnd, preStart + position - inStart + 1);
        node.left = left;
        node.right = right;
        return node;
    }

	public static void main(String[] args) {
		int[] preorder = {3, 1, 2, 5, 4};
		int[] inorder = {1, 2, 3, 4, 5};
		TreeNode root = buildTree(preorder, inorder);
		System.out.println(root.val);
		System.out.println(root.left.val);
		System.out.println(root.left.right.val);
		System.out.println(root.right.val);
		System.out.println(root.right.left.val);
	}

}
